src/cargo/sources/registry/http_remote.rs

```diff
@@ -0,0 +1,1209 @@
+                let handle = &pending[&token].1;
+                if let Some(result) = msg.result_for(handle) {
+                    results.push((token, result));
+                } else {
+                    debug!("message without a result (?)");
+                }
+            });
+
+            // Walk all the requests that completed and handle their responses.
+            //
+            // This will ultimately add more replies to self.downloads.eager, which we'll
+            while let Some((token, result)) = self.downloads.results.pop() {
+                trace!("{} finished with {:?}", token, result);
+
+                let (dl, handle) = self
+                    .downloads
+                    .pending
+                    .remove(&token)
+                    .expect("got a token for a non-in-progress transfer");
+
+                let data = dl.data.into_inner();
+                let mut handle = self.prefetch.remove(handle)?;
+                self.downloads.pending_ids.remove(&dl.path);
+
+                let fetched = Fetched {
+                    path: dl.path,
+                    name: dl.name,
+                    reqs: dl.reqs,
+                    is_transitive: dl.is_transitive,
+                };
+                assert!(
+                    self.fresh.insert(fetched.path.clone()),
+                    "downloaded the index file `{}` twice during prefetching",
+                    fetched.path.display(),
+                );
+
+                let code = handle.response_code()?;
+                debug!(
+                    "index file for {} downloaded with status code {}",
+                    fetched.name,
+                    handle.response_code()?
+                );
+
+                // This gets really noisy very quickly:
+                // self.config.shell().status("Prefetched", &fetched.name)?;
+
+                self.downloads.downloads_finished += 1;
+                self.downloads.downloaded_bytes += dl.total.get();
+                self.downloads.tick(WhyTick::DownloadFinished)?;
+
+                match code {
+                    200 => {
+                        // We got data back, hooray!
+                        // Let's update the index file.
+                        let path = self.config.assert_package_cache_locked(&self.index_path);
+                        let pkg = path.join(&fetched.path);
+                        paths::create_dir_all(pkg.parent().expect("pkg is a file"))?;
+                        let mut file = paths::create(pkg)?;
+                        file.write_all(dl.etag.into_inner().as_deref().unwrap_or("\n").as_bytes())?;
+                        file.write_all(
+                            dl.last_modified
+                                .into_inner()
+                                .as_deref()
+                                .unwrap_or("\n")
+                                .as_bytes(),
+                        )?;
+                        file.write_all(&data)?;
```

Comment:
I mentioned this briefly awhile back, but I think ideally we would never store the raw server data on disk. AFAIK there's no need for that, and instead we should be storing only the binary lookup format that's for the parsed summaries.

Would it be possible to do the transformation here so we don't have to reread this data later and process it?

I think also additionally we'd cache the data on disk but still retain it in-memory since we'll otherwise just later simply read it back from the disk.

Reply:
Unfortunately this ties into the broader point that the current source/registry API does not lend itself well to "partial invalidations". `RegistryIndex` discards the entire summary cache if the registry changes _at all_, which means it re-calls `load` and parses again even if that particular index file has not changed. To fix this, I think we'd have to have `load` be able to return something like `NotModified` rather than the actual data to tell `RegistryIndex` to continue using its previous parsed file. From a glance at that code, that will probably be a somewhat annoying change, and I think it's worth doing that separately from this PR.

As for keeping the bytes in memory, my guess is that the win will be marginal since the kernel's buffer cache almost certainly has the bytes still in memory when `load` is called. I think I would want to see benchmarks demonstrating a slow-down before adding an additional cache (that then also has to be maintained) here.

Reply:
Oh the index traits and such will definitely need some refactoring to fully support the http index and all its fidelity, I'm just sort of taking that as a given. Since this is just a performance optimization it seems reasonable to have a follow-up for it.

For keeping the bytes in-memory, I'm not really worried too much about perf but it seems like it's an extra line or two to just go ahead and populate an internal cache then we should do so. This depends on how the code is factored though to see how hard it is to just go ahead and populate the cache.